packages/utils/vue/props/runtime.ts

```diff
@@ -1,122 +1,47 @@
-
-  const _validator =
-    values || validator
-      ? (val: unknown) => {
-          let valid = false
-          let allowedValues: unknown[] = []
-
-          if (values) {
-            allowedValues = Array.from(values)
-            if (hasOwn(prop, 'default')) {
-              allowedValues.push(defaultValue)
-            }
-            valid ||= allowedValues.includes(val)
-          }
-          if (validator) valid ||= validator(val)
-
-          if (!valid && allowedValues.length > 0) {
-            const allowValuesText = [...new Set(allowedValues)]
-              .map((value) => JSON.stringify(value))
-              .join(', ')
-            warn(
-              `Invalid prop: validation failed${
-                key ? ` for prop "${key}"` : ''
-              }. Expected one of [${allowValuesText}], got value ${JSON.stringify(
-                val
-              )}.`
-            )
-          }
-          return valid
-        }
-      : undefined
-
-  const epProp: any = {
-    type,
-    required: !!required,
-    validator: _validator,
-    [epPropKey]: true,
-  }
-  if (hasOwn(prop, 'default')) epProp.default = defaultValue
-  return epProp
-}
-
-export const buildProps = <
-  Props extends Record<
-    string,
-    | { [epPropKey]: true }
-    | NativePropType
-    | EpPropInput<any, any, any, any, any>
-  >
->(
-  props: Props
-): {
-  [K in keyof Props]: IfEpProp<
-    Props[K],
-    Props[K],
-    IfNativePropType<Props[K], Props[K], EpPropConvert<Props[K]>>
-  >
-} =>
-  fromPairs(
-    Object.entries(props).map(([key, option]) => [
-      key,
-      buildProp(option as any, key),
-    ])
-  ) as any
+// eslint-disable-next-line prettier/prettier
+export const buildProp = <P extends EpProp<unknown, V, D>, const V extends Value<unknown>, const D extends Default<unknown, V>>(
+  prop: P & ConstraintP<P>
+) => prop as unknown as PruneProp<P>
+
+export const buildProps = <PP extends Record<string, EpProp<unknown, V, D | DA | DefaultFactory<T, V>>>, const T extends string | number | boolean | unknown[], const V extends Value<unknown>, const D extends Default<unknown, V>, DA extends Default<unknown[], V>>(props: PP & ConstraintPP<PP>) => { 
+  const entries =  Object.entries(props).map(([key, prop]) => {
+    if (!prop || isFunction(prop) || isArray(prop) || !prop.values) return [key, prop]
+
+    const { values, validator } = prop
+    const _validator = (val: T) => {
+      let valid = false
+      let allowValuesText = ''
+
+      valid ||= values.includes(val)
+      if (!valid) {
+        allowValuesText = values.map((value) => JSON.stringify(value)).join(', ')
+      }
+      if (validator) valid ||= validator(val)
+
+      if (!valid && allowValuesText) {
+        warn(
+          `Invalid prop: validation failed${
+            key ? ` for prop "${key}"` : ''
+          }. Expected one of [${allowValuesText}], got value ${JSON.stringify(
+            val
+          )}.`
+        )
```

Comment:
I notice that you moved the conversion of `values` from the `buildProp` function to the `buildProps` function.

I'm worried that someone will use the buildProp function directly in a component.

Reply:
If calling `buildProp` directly can also create a validator, the console warning might show which prop validation failed incorrectly, because its return value can be assigned to any prop. Since `buildProp` could give the impression that it's for creating validators, how about renaming it to `inferProp` or something?